In [100]:
import pandas as pd

### Reading in the data

In [101]:
data_2014 = pd.read_csv('Analytics_mindset_case_studies_PCard_FY2014.csv')

In [102]:
data_2015 = pd.read_csv('Analytics_mindset_case_studies_PCard_FY2015.csv')

/var/folders/f_/mkv__l410wg2mhmdf36mb_9c0000gn/T/ipykernel_35572/1353983471.py:1: DtypeWarning: Columns (7,8) have mixed types. Specify dtype option on import or set low_memory=False.
  data_2015 = pd.read_csv('Analytics_mindset_case_studies_PCard_FY2015.csv')


### Looking at the data

In [103]:
data_2014.head()

Agency Number                Agency Name Cardholder Last Name  \
0           1000  OKLAHOMA STATE UNIVERSITY                Mason   
1           1000  OKLAHOMA STATE UNIVERSITY                Mason   
2           1000  OKLAHOMA STATE UNIVERSITY               Massey   
3           1000  OKLAHOMA STATE UNIVERSITY               Massey   
4           1000  OKLAHOMA STATE UNIVERSITY        Mauro-Herrera   

  Cardholder First Initial                   Description  Amount  \
0                        C              GENERAL PURCHASE  890.00   
1                        C                  ROOM CHARGES  368.96   
2                        J              GENERAL PURCHASE  165.82   
3                        T              GENERAL PURCHASE   96.39   
4                        M  HAMMERMILL COPY PLUS COPY EA  125.96   

           Vendor Transaction Date     Posted Date  \
0           NACAS   7/30/2013 0:00  7/31/2013 0:00   
1  SHERATON HOTEL   7/30/2013 0:00  7/31/2013 0:00   
2  SEARS.COM 9300   7/29/2013 0:00  7/31/2013 0:00   
3  WAL-MART #0137   7/30/2013 0:00  7/31/2013 0:00   
4  STAPLES DIRECT   7/30/2013 0:00  7/31/2013 0:00   

                        Merchant Category Code (MCC)  
0        CHARITABLE AND SOCIAL SERVICE ORGANIZATIONS  
1                                           SHERATON  
2  DIRCT MARKETING/DIRCT MARKETERS--NOT ELSEWHERE...  
3                    GROCERY STORES,AND SUPERMARKETS  
4  STATIONERY, OFFICE SUPPLIES, PRINTING AND WRIT...

In [104]:
data_2015.head()

Agency Number                Agency Name Cardholder Last Name  \
0           1000  OKLAHOMA STATE UNIVERSITY                Clark   
1           1000  OKLAHOMA STATE UNIVERSITY                Clark   
2           1000  OKLAHOMA STATE UNIVERSITY                Clark   
3           1000  OKLAHOMA STATE UNIVERSITY                Clark   
4           1000  OKLAHOMA STATE UNIVERSITY                Clark   

  Cardholder First Initial                                        Description  \
0                        K                                   GENERAL PURCHASE   
1                        M                                   GENERAL PURCHASE   
2                        M                              NEWS ADV./SUBSC. EACH   
3                        M                                   GENERAL PURCHASE   
4                        P  BH 3-INX7-IN SMALL BRASS B PCS|BLUE HAWK SHOE ...   

       Amount                  Vendor Transaction Date     Posted Date  \
0     $88.88   Discount School Supply   7/23/2014 0:00  7/24/2014 0:00   
1    $425.00   BROONER'S FLOOR CENTER   7/23/2014 0:00  7/24/2014 0:00   
2    $141.50           820 NEWS PRESS   7/23/2014 0:00  7/24/2014 0:00   
3  $2,679.99           WWW.NEWEGG.COM   7/23/2014 0:00  7/24/2014 0:00   
4     $22.44             LOWES #00241   7/22/2014 0:00  7/24/2014 0:00   

                 Merchant Category Code (MCC)  
0  STATIONERY,OFFICE AND SCHOOL SUPPLY STORES  
1        FLOOR COVERING,RUG AND CARPET STORES  
2           CONTINUITY/SUBSCRIPTION MERCHANTS  
3                          ELECTRONICS STORES  
4                HOME SUPPLY WAREHOUSE STORES

In [105]:
data_2014.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 442458 entries, 0 to 442457
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Agency Number                 442458 non-null  int64  
 1   Agency Name                   442458 non-null  object 
 2   Cardholder Last Name          442458 non-null  object 
 3   Cardholder First Initial      442458 non-null  object 
 4   Description                   442458 non-null  object 
 5   Amount                        442458 non-null  float64
 6   Vendor                        442458 non-null  object 
 7   Transaction Date              442458 non-null  object 
 8   Posted Date                   442458 non-null  object 
 9   Merchant Category Code (MCC)  442458 non-null  object 
dtypes: float64(1), int64(1), object(8)
memory usage: 33.8+ MB


In [106]:
data_2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 437579 entries, 0 to 437578
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype 
---  ------                        --------------   ----- 
 0   Agency Number                 437579 non-null  int64 
 1   Agency Name                   437579 non-null  object
 2   Cardholder Last Name          437579 non-null  object
 3   Cardholder First Initial      437579 non-null  object
 4   Description                   437579 non-null  object
 5   Amount                        437579 non-null  object
 6   Vendor                        437579 non-null  object
 7   Transaction Date              214698 non-null  object
 8   Posted Date                   214698 non-null  object
 9   Merchant Category Code (MCC)  437579 non-null  object
dtypes: int64(1), object(9)
memory usage: 33.4+ MB


#### Notes:
- The amounts in 2015 have the dollar symbol while 2014 is classified as float, also they have commas and "()"
- Lost of nulls in Transaction Date and posted Date in 2015
- Column names have spaces

### Transforming Data in original Files

In [107]:
# Converting the Amount column to float
data_2015['Amount'] = data_2015['Amount'].str.replace('[\$,)]', '', regex=True)

# Changing (5.15) to -5.15
data_2015['Amount'] = data_2015['Amount'].str.replace('(', '-', regex = True)

data_2015['Amount'] = data_2015['Amount'].astype(float)

In [108]:
data_2015.dtypes

Agency Number                     int64
Agency Name                      object
Cardholder Last Name             object
Cardholder First Initial         object
Description                      object
Amount                          float64
Vendor                           object
Transaction Date                 object
Posted Date                      object
Merchant Category Code (MCC)     object
dtype: object

In [109]:
# Removing WhiteSpaces from the column names before merging -> as per instructions

# remove spaces from column names of data_2014
data_2014.rename(columns=lambda x: x.replace(' ', ''), inplace=True)

# remove spaces from column names of data_2015
data_2015.rename(columns=lambda x: x.replace(' ', ''), inplace=True)

In [110]:
dataset = pd.concat([data_2014, data_2015]).reset_index(drop=True)

In [111]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 880037 entries, 0 to 880036
Data columns (total 10 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   AgencyNumber               880037 non-null  int64  
 1   AgencyName                 880037 non-null  object 
 2   CardholderLastName         880037 non-null  object 
 3   CardholderFirstInitial     880037 non-null  object 
 4   Description                880037 non-null  object 
 5   Amount                     880037 non-null  float64
 6   Vendor                     880037 non-null  object 
 7   TransactionDate            657156 non-null  object 
 8   PostedDate                 657156 non-null  object 
 9   MerchantCategoryCode(MCC)  880037 non-null  object 
dtypes: float64(1), int64(1), object(8)
memory usage: 67.1+ MB


In [112]:
dataset['TransactionDate'] = pd.to_datetime(dataset['TransactionDate'], format='%m/%d/%Y %H:%M')
dataset['PostedDate'] = pd.to_datetime(dataset['PostedDate'], format='%m/%d/%Y %H:%M')

# All times are 0:00, remove all of them
# Not sure if I should set NaT's to ''

In [113]:
dataset['MerchantCategoryCode(MCC)'] = dataset['MerchantCategoryCode(MCC)'].str.lower()

In [114]:
# Removing trailing and leading spaces from data
dataset = dataset.apply(lambda x: x.str.strip() if x.dtype == "object" else x)


In [115]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 880037 entries, 0 to 880036
Data columns (total 10 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   AgencyNumber               880037 non-null  int64         
 1   AgencyName                 880037 non-null  object        
 2   CardholderLastName         880037 non-null  object        
 3   CardholderFirstInitial     880037 non-null  object        
 4   Description                880037 non-null  object        
 5   Amount                     880037 non-null  float64       
 6   Vendor                     880037 non-null  object        
 7   TransactionDate            657156 non-null  datetime64[ns]
 8   PostedDate                 657156 non-null  datetime64[ns]
 9   MerchantCategoryCode(MCC)  880037 non-null  object        
dtypes: datetime64[ns](2), float64(1), int64(1), object(6)
memory usage: 67.1+ MB


In [116]:
dataset.head()

AgencyNumber                 AgencyName CardholderLastName  \
0          1000  OKLAHOMA STATE UNIVERSITY              Mason   
1          1000  OKLAHOMA STATE UNIVERSITY              Mason   
2          1000  OKLAHOMA STATE UNIVERSITY             Massey   
3          1000  OKLAHOMA STATE UNIVERSITY             Massey   
4          1000  OKLAHOMA STATE UNIVERSITY      Mauro-Herrera   

  CardholderFirstInitial                   Description  Amount  \
0                      C              GENERAL PURCHASE  890.00   
1                      C                  ROOM CHARGES  368.96   
2                      J              GENERAL PURCHASE  165.82   
3                      T              GENERAL PURCHASE   96.39   
4                      M  HAMMERMILL COPY PLUS COPY EA  125.96   

           Vendor TransactionDate PostedDate  \
0           NACAS      2013-07-30 2013-07-31   
1  SHERATON HOTEL      2013-07-30 2013-07-31   
2  SEARS.COM 9300      2013-07-29 2013-07-31   
3  WAL-MART #0137      2013-07-30 2013-07-31   
4  STAPLES DIRECT      2013-07-30 2013-07-31   

                           MerchantCategoryCode(MCC)  
0        charitable and social service organizations  
1                                           sheraton  
2  dirct marketing/dirct marketers--not elsewhere...  
3                    grocery stores,and supermarkets  
4  stationery, office supplies, printing and writ...

In [117]:
# dataset.to_csv('task3_final_dataset.csv', index = False)

### ANSWERS

##### There are 880037 rows in the final file
##### The total sum of the amount column is 379,101,777.6800001

In [118]:
dataset['Amount'].sum()

# Double checked in excel and the sum is correct

379101777.6800001

In [119]:
dataset.shape

(880037, 10)

In [120]:
dataset.shape

(880037, 10)

In [121]:
#dataset for OSU 2014
o_data=dataset[(dataset["AgencyName"]=="OKLAHOMA STATE UNIVERSITY") & (dataset['TransactionDate'].dt.strftime('%Y')=='2014')]
o_data

AgencyNumber                 AgencyName CardholderLastName  \
56333           1000  OKLAHOMA STATE UNIVERSITY         Reddington   
56334           1000  OKLAHOMA STATE UNIVERSITY              Reece   
56335           1000  OKLAHOMA STATE UNIVERSITY              Reece   
56336           1000  OKLAHOMA STATE UNIVERSITY            Rendina   
56337           1000  OKLAHOMA STATE UNIVERSITY                Rex   
...              ...                        ...                ...   
581693          1000  OKLAHOMA STATE UNIVERSITY            Lindsey   
581694          1000  OKLAHOMA STATE UNIVERSITY              Lloyd   
581695          1000  OKLAHOMA STATE UNIVERSITY             Longan   
581696          1000  OKLAHOMA STATE UNIVERSITY              Lopez   
581697          1000  OKLAHOMA STATE UNIVERSITY              Fokar   

       CardholderFirstInitial  \
56333                       D   
56334                       S   
56335                       S   
56336                       E   
56337                       T   
...                       ...   
581693                      K   
581694                      J   
581695                      P   
581696                      R   
581697                      M   

                                              Description  Amount  \
56333                                    GENERAL PURCHASE  735.00   
56334                                    GENERAL PURCHASE  304.04   
56335                                    GENERAL PURCHASE   39.19   
56336                                    GENERAL PURCHASE  260.00   
56337                                    GENERAL PURCHASE   80.00   
...                                                   ...     ...   
581693                                   GENERAL PURCHASE   15.00   
581694                                   GENERAL PURCHASE   93.21   
581695                                   GENERAL PURCHASE    2.80   
581696         12-OZ COLOR RAD CARAMEL VA EA|#72 CLAMP EA   17.72   
581697  CELLYTIC B 2X EA|PROTEASE INHIBITOR COCKTA EA|...  701.54   

                           Vendor TransactionDate PostedDate  \
56333                        ABSA      2014-01-03 2014-01-06   
56334          ORSCHELN DURANT 83      2014-01-03 2014-01-06   
56335          ORSCHELN DURANT 83      2014-01-03 2014-01-06   
56336              CELLSIGNAL.COM      2014-01-02 2014-01-06   
56337   THERMACUBE       00 OF 00      2014-01-02 2014-01-06   
...                           ...             ...        ...   
581693     OK STATE BUREAU OF INV      2014-12-01 2014-12-02   
581694         BUFFALO WILD WINGS      2014-12-01 2014-12-02   
581695           PIKEPASS REBILLS      2014-12-01 2014-12-02   
581696               LOWES #00241      2014-12-01 2014-12-02   
581697           SIGMA ALDRICH US      2014-12-01 2014-12-02   

                                MerchantCategoryCode(MCC)  
56333   schools and educational services not elsewhere...  
56334                                     hardware stores  
56335                                     hardware stores  
56336   drugs, drug proprietaries, and druggists sundries  
56337          business services not elsewhere classified  
...                                                   ...  
581693      government services--not elsewhere classified  
581694                      eating places and restaurants  
581695                              tolls and bridge fees  
581696                       home supply warehouse stores  
581697  dirct marketing/dirct marketers--not elsewhere...  

[116031 rows x 10 columns]

1. User shall not spend more than $5,000 per transaction.

In [122]:
#Transactions in 2014 that was for more than $5,000
exceed_limit= o_data[o_data["Amount"]>5000]
trasaction_details = exceed_limit[["Amount", "CardholderLastName", "CardholderFirstInitial", "Description", "Vendor", "TransactionDate", "PostedDate", "MerchantCategoryCode(MCC)"]]
exceed_limit_transaction_detail = trasaction_details.sort_values(by="Amount", ascending=False)
exceed_limit_transaction_detail

Amount CardholderLastName CardholderFirstInitial       Description  \
453178  29731.61              Clark                      C  GENERAL PURCHASE   
375945  27657.55             Ropers                      A      ROOM CHARGES   
369337  22764.84             Ropers                      A      ROOM CHARGES   
468012  21243.93             Ropers                      A  GENERAL PURCHASE   
579359  19233.28             Stover                      C  GENERAL PURCHASE   
468440  16517.20             Ropers                      A        AIR TRAVEL   
463410  15327.36           Williams                      J      ROOM CHARGES   
454866  13186.64           Bilbeisi                      S      ROOM CHARGES   
60502   12600.00              Clark                      C  GENERAL PURCHASE   
276912  12380.00              Clark                      C  GENERAL PURCHASE   
443270  11944.42             Ropers                      A  GENERAL PURCHASE   
372592  11643.75             Stover                      C      ROOM CHARGES   
283657  10299.66             Stover                      C      ROOM CHARGES   
267581  10153.35             Stover                      C      ROOM CHARGES   
265210   9770.40             Ropers                      A  GENERAL PURCHASE   
269706   7951.00             Stover                      C      ROOM CHARGES   
566017   7189.81             Stover                      C      ROOM CHARGES   
559304   7013.02             Ropers                      A  GENERAL PURCHASE   
278895   6980.00             Ropers                      A  GENERAL PURCHASE   
258267   6923.32             Stover                      C      ROOM CHARGES   
570424   6905.52             Stover                      C      ROOM CHARGES   
277282   6545.02             Ropers                      A      ROOM CHARGES   
283457   6400.87             Stover                      C      ROOM CHARGES   
280454   6266.53             Ropers                      A  GENERAL PURCHASE   
275929   6210.50             Ropers                      A  GENERAL PURCHASE   
381206   6000.00              Clark                      C  GENERAL PURCHASE   
452658   5990.29             Ropers                      A  GENERAL PURCHASE   
577594   5758.35             Stover                      C  GENERAL PURCHASE   
270368   5731.60             Stover                      C      ROOM CHARGES   
256035   5688.46             Stover                      C  GENERAL PURCHASE   
368043   5638.18              Krout                      K  GENERAL PURCHASE   
274400   5123.25             Stover                      C      ROOM CHARGES   
267857   5101.20             Stover                      C      ROOM CHARGES   

                           Vendor TransactionDate PostedDate  \
453178       OEC OTIS ELEVATOR CO      2014-08-18 2014-08-19   
375945          WINGATE MANHATTAN      2014-05-22 2014-05-23   
369337   HAMPTON INN & SUITES CRN      2014-05-23 2014-05-26   
468012          MAGDALENE COLLEGE      2014-09-08 2014-09-09   
579359    COCKRELL EYECARE CENTER      2014-12-24 2014-12-26   
468440  SOUTHWES    5262393866444      2014-08-28 2014-09-01   
463410     HILTON BOSTON DOWNTOWN      2014-09-24 2014-09-26   
454866                ROWN NYC FD      2014-08-01 2014-08-04   
60502                TLC PAINTING      2014-01-24 2014-01-27   
276912     AABB-AMER. ASSO. BLOOD      2014-04-04 2014-04-07   
443270     NASA CATERING 10550515      2014-07-23 2014-07-24   
372592     EMBASSY SUITES CENTRAL      2014-04-30 2014-05-01   
283657     WATERFRONT PLACE HOTEL      2014-04-13 2014-04-21   
267581  COURTYARD BY MARRIOTT ARL      2014-03-20 2014-03-24   
265210         TOPGOLF COLONY F&B      2014-02-28 2014-03-03   
269706   HOLIDAY INN EXPRESS & SU      2014-03-06 2014-03-10   
566017            SHERATON DALLAS      2014-11-05 2014-11-07   
559304          MAGDALENE COLLEGE      2014-10-21 2014-10-22   
278895            ASSETWORKS INC.      2014-04-02 2014-04-03   
258267  C

In [147]:
# count number of rows
len(exceed_limit_transaction_detail)

33

2. User shall not spend more than $50,000 per year.

In [123]:
#transactions grouped by user
transaction_user=o_data.groupby(["CardholderLastName","CardholderFirstInitial"])["Amount"].sum().reset_index(name ='YearAmount')

In [124]:
#Exceeds the yearly limt
Year_limit= transaction_user[transaction_user["YearAmount"]>50000].sort_values(by="YearAmount", ascending=False)
Year_limit

CardholderLastName CardholderFirstInitial  YearAmount
788               Hines                      G  1595302.32
1806          Tornakian                      M  1182094.86
779              Heusel                      J   627608.56
579         Fitzpatrick                      S   437479.84
194              Bowers                      R   392761.34
...                 ...                    ...         ...
707           Gruntmeir                      J    50553.45
855         Huffstetler                      N    50352.46
1975           Woodward                      K    50318.47
1950           Williams                      G    50297.55
1737          Sundstrom                      K    50234.57

[127 rows x 3 columns]

3. User shall not spend more
than $10,000 per month
without approval.

In [125]:
# extract month information from transaction date
o_data['Month'] = o_data['TransactionDate'].dt.strftime('%m')
o_data.head()

/var/folders/f_/mkv__l410wg2mhmdf36mb_9c0000gn/T/ipykernel_35572/494233787.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  o_data['Month'] = o_data['TransactionDate'].dt.strftime('%m')


AgencyNumber                 AgencyName CardholderLastName  \
56333          1000  OKLAHOMA STATE UNIVERSITY         Reddington   
56334          1000  OKLAHOMA STATE UNIVERSITY              Reece   
56335          1000  OKLAHOMA STATE UNIVERSITY              Reece   
56336          1000  OKLAHOMA STATE UNIVERSITY            Rendina   
56337          1000  OKLAHOMA STATE UNIVERSITY                Rex   

      CardholderFirstInitial       Description  Amount  \
56333                      D  GENERAL PURCHASE  735.00   
56334                      S  GENERAL PURCHASE  304.04   
56335                      S  GENERAL PURCHASE   39.19   
56336                      E  GENERAL PURCHASE  260.00   
56337                      T  GENERAL PURCHASE   80.00   

                          Vendor TransactionDate PostedDate  \
56333                       ABSA      2014-01-03 2014-01-06   
56334         ORSCHELN DURANT 83      2014-01-03 2014-01-06   
56335         ORSCHELN DURANT 83      2014-01-03 2014-01-06   
56336             CELLSIGNAL.COM      2014-01-02 2014-01-06   
56337  THERMACUBE       00 OF 00      2014-01-02 2014-01-06   

                               MerchantCategoryCode(MCC) Month  
56333  schools and educational services not elsewhere...    01  
56334                                    hardware stores    01  
56335                                    hardware stores    01  
56336  drugs, drug proprietaries, and druggists sundries    01  
56337         business services not elsewhere classified    01

In [126]:
#transactions grouped by user and month
transaction_user_month=o_data.groupby(["CardholderLastName","CardholderFirstInitial", "Month"])["Amount"].sum().reset_index(name ='MonthlyAmount')
transaction_user_month.head()

CardholderLastName CardholderFirstInitial Month  MonthlyAmount
0             Abbott                      T    01        2149.72
1             Abbott                      T    02        2443.81
2             Abbott                      T    03        3874.77
3             Abbott                      T    04        4183.66
4             Abbott                      T    05        9166.46

In [127]:
#Exceeds the monthly limit
#sort by month and monthly amount
month_limit= transaction_user_month[transaction_user_month["MonthlyAmount"]>10000].sort_values(by=["Month", "MonthlyAmount"], ascending=[True, False])
month_limit


CardholderLastName CardholderFirstInitial Month  MonthlyAmount
6110               Hines                      G    01      149700.14
13827          Tornakian                      M    01      112598.39
6051              Heusel                      J    01      110190.56
4526         Fitzpatrick                      S    01       72128.93
1585              Bowers                      R    01       37935.56
...                  ...                    ...   ...            ...
8307                Lowe                      K    12       11348.34
8499             Machart                      D    12       10626.26
7154             KELOUGH                      N    12       10253.47
836              Ballard                      J    12       10244.69
13300          Sundstrom                      K    12       10221.92

[457 rows x 4 columns]

4. An amount more than
$5,000 should not be split
between two or more swipes
of the card by the same
person.

In [128]:
# filter out the columns we want
filter_data = o_data[["CardholderLastName","CardholderFirstInitial", "Vendor", "TransactionDate", "Amount"]]

In [129]:
# keep those with repeated entries
# in other words, transactions for two or more swipes
repeated_entries = filter_data[filter_data.duplicated(subset=["CardholderLastName","CardholderFirstInitial", "Vendor", "TransactionDate"], keep=False)]
repeated_entries

CardholderLastName CardholderFirstInitial                  Vendor  \
56334               Reece                      S      ORSCHELN DURANT 83   
56335               Reece                      S      ORSCHELN DURANT 83   
56342            Robinson                      S     INTEGRATED DNA TECH   
56343            Robinson                      S     INTEGRATED DNA TECH   
56344            Robinson                      S     INTEGRATED DNA TECH   
...                   ...                    ...                     ...   
581666             Graham                      O          WED LINK MEDIA   
581667             Graham                      O          WED LINK MEDIA   
581668             Graham                      O        HIBU INC. - WEST   
581679              Hines                      G  STILLWATER WINNELSON C   
581680              Hines                      G  STILLWATER WINNELSON C   

       TransactionDate  Amount  
56334       2014-01-03  304.04  
56335       2014-01-03   39.19  
56342       2014-01-03   31.00  
56343       2014-01-03  111.75  
56344       2014-01-03   49.50  
...                ...     ...  
581666      2014-12-01  300.00  
581667      2014-12-01  313.95  
581668      2014-12-01   92.00  
581679      2014-11-24    1.90  
581680      2014-11-24  383.20  

[27079 rows x 5 columns]

In [130]:
# sum up for each same-day splitted entries
repeated_daily_entries=repeated_entries.groupby(["CardholderLastName","CardholderFirstInitial", "Vendor", "TransactionDate"])["Amount"].sum().reset_index(name ='DailyAmount')

In [131]:
# Filter those with >5000 daily limits; sort by transaction dates
violate_with_split = repeated_daily_entries[repeated_daily_entries["DailyAmount"]>5000].sort_values(by="TransactionDate", ascending=True)
violate_with_split

CardholderLastName CardholderFirstInitial                     Vendor  \
9210           Tornakian                      M         R.B. AKINS COMPANY   
9063           Tornakian                      M  AUTOMATED BUILDNG SYSTEMS   
4854               Hines                      G   OKLAHOMA CONTRACTORS SUP   
4056              Heusel                      J             ANIXTER-115687   
5679             Kennedy                      J   COUNCIL OF GRADUATE SCHO   
...                  ...                    ...                        ...   
9198           Tornakian                      M     OKLAHOMA CITY WINNELSO   
10142               Wood                      C   MEDLEY MATERIAL HANDLING   
9979           Wilkinson                      T         PHONAK HEARING SYS   
5785            Kindschi                      J         VETERANS AFFRS DMC   
2995             Edwards                      W   IN  MAVERICK JACKETS INC   

      TransactionDate  DailyAmount  
9210       2014-01-06      5164.00  
9063       2014-01-08      6063.30  
4854       2014-01-16      5130.00  
4056       2014-01-23      5775.10  
5679       2014-02-04      5166.00  
...               ...          ...  
9198       2014-12-04      5966.80  
10142      2014-12-12      5296.32  
9979       2014-12-17      6214.60  
5785       2014-12-22      6582.50  
2995       2014-12-24      5398.35  

[66 rows x 5 columns]

5. Continued from 4. Count how often each individual purchaser did not follow the
previous restriction. Sort this in descending order by the count

In [132]:
# count the number of violations for each purchaser
count_split_violation = violate_with_split.groupby(["CardholderLastName","CardholderFirstInitial"])["DailyAmount"].count().reset_index(name ='Violations')

In [133]:
# sort with ascending order
count_split_violation.sort_values(by="Violations", ascending=False)

CardholderLastName CardholderFirstInitial  Violations
14              Hines                      G          18
24          Tornakian                      M          11
16           Kindschi                      J           8
13             Heusel                      J           2
11            Gotcher                      M           2
21   Santhanakrishnan                      A           1
18               Lowe                      K           1
19              Morey                      A           1
20             Munday                      T           1
0              Bailey                      J           1
17             Lenoir                      K           1
23             Stover                      C           1
25           Whitmore                      D           1
26               Wick                      J           1
27          Wilkinson                      T           1
28           Williams                      J           1
22              Singh                      R           1
15            Kennedy                      J           1
1               Baker                      B           1
12             Heaton                      D           1
10            Forquer                      H           1
9          Eisensmith                      J           1
8             Edwards                      W           1
7               Davis                      E           1
6              Cowell                      C           1
5              Cooper                      E           1
4               Clark                      M           1
3               Clark                      C           1
2             Brinker                      B           1
29               Wood                      C           1

6. Purchases should not be
split between two or more
cardholders.

In [134]:
filter_data.head() #contains the columns we want


CardholderLastName CardholderFirstInitial                     Vendor  \
56333         Reddington                      D                       ABSA   
56334              Reece                      S         ORSCHELN DURANT 83   
56335              Reece                      S         ORSCHELN DURANT 83   
56336            Rendina                      E             CELLSIGNAL.COM   
56337                Rex                      T  THERMACUBE       00 OF 00   

      TransactionDate  Amount  
56333      2014-01-03  735.00  
56334      2014-01-03  304.04  
56335      2014-01-03   39.19  
56336      2014-01-02  260.00  
56337      2014-01-02   80.00

In [151]:
 # contains same name, same vendor, same date transactions
repeated_entries

CardholderLastName CardholderFirstInitial                  Vendor  \
56334               Reece                      S      ORSCHELN DURANT 83   
56335               Reece                      S      ORSCHELN DURANT 83   
56342            Robinson                      S     INTEGRATED DNA TECH   
56343            Robinson                      S     INTEGRATED DNA TECH   
56344            Robinson                      S     INTEGRATED DNA TECH   
...                   ...                    ...                     ...   
581666             Graham                      O          WED LINK MEDIA   
581667             Graham                      O          WED LINK MEDIA   
581668             Graham                      O        HIBU INC. - WEST   
581679              Hines                      G  STILLWATER WINNELSON C   
581680              Hines                      G  STILLWATER WINNELSON C   

       TransactionDate  Amount  
56334       2014-01-03  304.04  
56335       2014-01-03   39.19  
56342       2014-01-03   31.00  
56343       2014-01-03  111.75  
56344       2014-01-03   49.50  
...                ...     ...  
581666      2014-12-01  300.00  
581667      2014-12-01  313.95  
581668      2014-12-01   92.00  
581679      2014-11-24    1.90  
581680      2014-11-24  383.20  

[27079 rows x 5 columns]

In [150]:
# contains all same vendor, same date transactions
repeated_vendors = filter_data[filter_data.duplicated(subset=["Vendor", "TransactionDate"], keep=False)]   
repeated_vendors

CardholderLastName CardholderFirstInitial                     Vendor  \
56334               Reece                      S         ORSCHELN DURANT 83   
56335               Reece                      S         ORSCHELN DURANT 83   
56340             Roberts                      C    AIR LIQUIDE AM SPEC GAS   
56342            Robinson                      S        INTEGRATED DNA TECH   
56343            Robinson                      S        INTEGRATED DNA TECH   
...                   ...                    ...                        ...   
581690        Kithakye CH                      S             WAL-MART #0137   
581691           Knobloch                      M  ALLEGRA PRINT AND IMAGING   
581695             Longan                      P           PIKEPASS REBILLS   
581696              Lopez                      R               LOWES #00241   
581697              Fokar                      M           SIGMA ALDRICH US   

       TransactionDate   Amount  
56334       2014-01-03   304.04  
56335       2014-01-03    39.19  
56340       2014-01-03   132.00  
56342       2014-01-03    31.00  
56343       2014-01-03   111.75  
...                ...      ...  
581690      2014-12-01    55.57  
581691      2014-12-01  1559.45  
581695      2014-12-01     2.80  
581696      2014-12-01    17.72  
581697      2014-12-01   701.54  

[62807 rows x 5 columns]

In [152]:
# A split purchasing case is either splitted by (1) multiple swipes by one cardholder or (2) different cardholders
# Subtract (1) from all split purchasing orders, we expect to get (2)
merged = pd.merge(repeated_entries[["Vendor", "TransactionDate", "Amount"]], 
                  repeated_vendors[["Vendor", "TransactionDate", "Amount"]],
                    on=["Vendor", "TransactionDate"], how='outer',  suffixes=['_1', '_2'], indicator=True)
vendor_diff_cardholders = merged.loc[merged['_merge'] == 'right_only', ["Vendor", "TransactionDate", "Amount_2"]].rename(columns={'Amount_2': 'Amount'})
vendor_diff_cardholders

Vendor TransactionDate  Amount
180414  AIR LIQUIDE AM SPEC GAS      2014-01-03  132.00
180415  AIR LIQUIDE AM SPEC GAS      2014-01-03   16.43
180416  AIR LIQUIDE AM SPEC GAS      2014-01-03  353.20
180417           WAL-MART #0137      2014-01-06   16.97
180418           WAL-MART #0137      2014-01-06   51.91
...                         ...             ...     ...
202735     EUREKA WATER COMPANY      2014-12-02   27.50
202736    STILLWATER WINLECTRIC      2014-12-01  411.05
202737    STILLWATER WINLECTRIC      2014-12-01  336.24
202738                  DROPBOX      2014-11-30    9.99
202739                  DROPBOX      2014-11-30    9.99

[22326 rows x 3 columns]

In [138]:
# sum up for each same-day vendor amount with different purchasers
vendor_diff_card_sum = vendor_diff_cardholders.groupby(["Vendor", "TransactionDate"])["Amount"].sum().reset_index(name ='VendorDailyAmount')
vendor_diff_card_sum

Vendor TransactionDate  VendorDailyAmount
0     182 BUILD-TULSA-BROKEN      2014-01-22            1451.46
1            3D ROBOTICS INC      2014-01-31             661.25
2                   4IMPRINT      2014-02-12            1769.79
3                   4IMPRINT      2014-03-28             814.99
4                   4IMPRINT      2014-04-15            2979.69
...                      ...             ...                ...
7381  YANG LING GUO JI HIU L      2014-09-21             526.52
7382  YANG LING GUO JI HIU L      2014-11-27             409.42
7383              ZAZZLE.COM      2014-05-23              60.80
7384  ZEE MEDICAL   00022012      2014-02-04             452.24
7385          ZORO TOOLS INC      2014-05-06             353.96

[7386 rows x 3 columns]

In [139]:
violate_with_diff_card = vendor_diff_card_sum[vendor_diff_card_sum["VendorDailyAmount"]>5000].sort_values(by="TransactionDate", ascending=True)
violate_with_diff_card

Vendor TransactionDate  VendorDailyAmount
1731                        EEI      2014-01-03            6128.60
2821     IVG INVITROGEN CORPORA      2014-01-07            5524.16
72                         AFLV      2014-01-10            6935.00
847       B & H PHOTO-VIDEO.COM      2014-02-13            5088.14
1732                        EEI      2014-02-14            6169.00
2842     IVG INVITROGEN CORPORA      2014-02-21            5557.35
1970       FENTON OFFICE SUPPLY      2014-03-14            5702.97
6088            THE LEE COMPANY      2014-03-19            5502.39
5708              STUART C IRBY      2014-03-31            6040.82
5469  STILLWATER SCREENPRINTING      2014-04-01            7794.00
441               ANASPEC, INC.      2014-04-08            6475.00
1976       FENTON OFFICE SUPPLY      2014-04-24            5118.24
5915     TFS FISHERSCI ECOM HUS      2014-05-03            5748.92
4245          PAYPAL  FUTUREBUS      2014-05-15            8029.80
4719      SHAKE RATTLE AND ROLL      2014-05-28            9938.00
4720      SHAKE RATTLE AND ROLL      2014-06-09            6813.00
1984       FENTON OFFICE SUPPLY      2014-06-10            7339.30
5940     TFS FISHERSCI ECOM HUS      2014-07-12            5223.65
3965      OK DEPT OF VO-TECH ED      2014-07-21            6285.00
5278         STERIS CORPORATION      2014-07-23            5071.85
7362      XEROX CORPORATION/RBO      2014-07-28            5118.32
6090            THE LEE COMPANY      2014-07-30            5651.36
707       AWL PEARSON EDUCATION      2014-08-12            5265.03
251   ALLEGRA PRINT AND IMAGING      2014-08-21            5090.51
3607   MWI VETERINARY SUPPLY CO      2014-08-27            8709.20
1998       FENTON OFFICE SUPPLY      2014-08-27            5587.31
4156            P & K EQUIPMENT      2014-09-18            5172.56
2005       FENTON OFFICE SUPPLY      2014-09-19            5766.88
1340     CHRIS' UNIVERSITY SPIR      2014-09-23            5374.50
3608   MWI VETERINARY SUPPLY CO      2014-09-24            5993.29
1762     ESKIMO JOES PROMO PROD      2014-09-25            5073.60
9                      4IMPRINT      2014-09-26            7313.16
4875     SIMPLEX GRINNELL WEB P      2014-09-30            6453.53
6349      VWR INTERNATIONAL INC      2014-10-02            5277.93
7279                WW GRAINGER      2014-10-06            5032.69
1483   CREEKSTONE FARMS PREMIUM      2014-10-14            5069.41
2939      KINNUNEN SALES & RENT      2014-10-22            5190.17
4376          PRINTING SERVICES      2014-11-06            6289.00
1756        EQUIPMENT AFFILIATE      2014-11-11            5714.00
4939   SQ  GRANNY HAD ONE, INC.      2014-12-12            6125.00
2872          JW LEES FURNITURE      2014-12-15            6760.00

In [149]:
# count number of violations
len(violate_with_diff_card)

41

7. Transactions are prohibited
for expenses for food and
mileage while traveling. A
per diem for food expenses
and mileage may be claimed
using a travel voucher.

In [140]:
# Extract traveling dates of all purchasers
travel_dates = o_data[o_data["MerchantCategoryCode(MCC)"].str.contains('|'.join(['hotel', 'motel', 'resort', 'inn']))]
travel_dates = travel_dates[["CardholderLastName","CardholderFirstInitial", "TransactionDate"]].drop_duplicates()
travel_dates.sort_values(by="TransactionDate")

CardholderLastName CardholderFirstInitial TransactionDate
57035              DeRaps                      M      2014-01-02
58232              Wooten                      C      2014-01-02
57284             Johnson                      S      2014-01-07
57941              Hoxsie                      A      2014-01-08
62199              Bailey                      J      2014-01-08
...                   ...                    ...             ...
578246            Wilkins                      M      2014-12-22
579927           Cartmell                      D      2014-12-23
578947           Conaghan                      J      2014-12-24
579356              Smith                      B      2014-12-25
579709             Duncan                      T      2014-12-27

[2085 rows x 3 columns]

In [141]:
# Extract all food related transactions
food_transactions = o_data[o_data["MerchantCategoryCode(MCC)"].str.contains('|'.join(['food', 'restaurant']))]
food_transactions.head()

AgencyNumber                 AgencyName CardholderLastName  \
56446          1000  OKLAHOMA STATE UNIVERSITY              Biggs   
56543          1000  OKLAHOMA STATE UNIVERSITY            Johnson   
56544          1000  OKLAHOMA STATE UNIVERSITY              Jones   
56545          1000  OKLAHOMA STATE UNIVERSITY              Jones   
56718          1000  OKLAHOMA STATE UNIVERSITY               Ward   

      CardholderFirstInitial       Description   Amount  \
56446                      J  GENERAL PURCHASE  1236.48   
56543                      S  GENERAL PURCHASE   131.94   
56544                      D  GENERAL PURCHASE   775.00   
56545                      D  GENERAL PURCHASE   230.00   
56718                      J  GENERAL PURCHASE   100.00   

                      Vendor TransactionDate PostedDate  \
56446   TEXAS ROADHOUSE 2299      2014-01-10 2014-01-10   
56543  CHEONGJIN DONG KOREAN      2014-01-05 2014-01-06   
56544  COBA SELECT SIRES INC      2014-01-02 2014-01-06   
56545  COBA SELECT SIRES INC      2014-01-02 2014-01-06   
56718       PURPLE SAGE CAFE      2014-01-06 2014-01-08   

             MerchantCategoryCode(MCC) Month  
56446    eating places and restaurants    01  
56543    eating places and restaurants    01  
56544  pet shops,pet food and supplies    01  
56545  pet shops,pet food and supplies    01  
56718            fast food restaurants    01

In [153]:
# Filter food transactions on those days with travelling
join = food_transactions.merge(travel_dates, how="inner", on=["CardholderLastName","CardholderFirstInitial","TransactionDate"])
food_violations = join[["Amount", "CardholderLastName", "CardholderFirstInitial", "Description", "Vendor", "TransactionDate", "PostedDate", "MerchantCategoryCode(MCC)"]
     ].sort_values(by=["CardholderLastName", "CardholderFirstInitial", "Amount"], ascending=[True, True,True])
food_violations

Amount CardholderLastName CardholderFirstInitial       Description  \
27    6.67              Baker                      B  GENERAL PURCHASE   
26   15.00              Baker                      B  GENERAL PURCHASE   
25   35.00              Baker                      B  GENERAL PURCHASE   
28  110.80              Baker                      B  GENERAL PURCHASE   
91  973.51              Biggs                      J  GENERAL PURCHASE   
..     ...                ...                    ...               ...   
19   20.70              Young                      C  GENERAL PURCHASE   
18   96.96              Young                      C  GENERAL PURCHASE   
17  145.85              Young                      C  GENERAL PURCHASE   
42  175.00              Young                      C  GENERAL PURCHASE   
36  270.82              Young                      C  GENERAL PURCHASE   

                    Vendor TransactionDate PostedDate  \
27      CHICK-FIL-A #01973      2014-03-08 2014-03-10   
26       JIMMY JOHNS # 563      2014-03-08 2014-03-10   
25    MCALISTER'S DELI 710      2014-03-08 2014-03-10   
28       JIMMY JOHNS # 563      2014-03-08 2014-03-10   
91  THE OLIVE GARD00044263      2014-11-07 2014-11-10   
..                     ...             ...        ...   
19       MCDONALD'S F28697      2014-03-30 2014-04-01   
18     PITA PIT - US - 035      2014-03-30 2014-04-01   
17   BOBBY T'S - MANHATTAN      2014-03-30 2014-03-31   
42     GENGHIS GRILL-MOORE      2014-04-19 2014-04-21   
36     BJS RESTAURANTS 471      2014-04-13 2014-04-14   

        MerchantCategoryCode(MCC)  
27          fast food restaurants  
26          fast food restaurants  
25          fast food restaurants  
28          fast food restaurants  
91  eating places and restaurants  
..                            ...  
19          fast food restaurants  
18          fast food restaurants  
17  eating places and restaurants  
42  eating places and restaurants  
36  eating places and restaurants  

[99 rows x 8 columns]

In [157]:
# go one step further, count how many people have violated
len(food_violations.groupby(["CardholderLastName", "CardholderFirstInitial"])["Amount"].count())

31